### PR TITLE
Superclustering DNN in TICLv5

### PR DESCRIPTION
**Pull request by tcuisset**
To test, you need to download the model (attached), rename it to `supercls_v2.pb`and move it to `CMSSW_14_1_X_2024-04-18-2300/external/el8_amd64_gcc12/data/RecoHGCal/TICL/data/tf_models/supercls_v2.pb`, replacing "el8_*" by your value of `$SCRAM_ARCH`
[supercls_v2.pb.txt](https://github.com/felicepantaleo/cmssw/files/14090080/supercls_v2.pb.txt)

### CODE REVIEW

File: RecoHGCal/TICL/plugins/EGammaSuperclusterProducer.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    reco::CaloCluster& caloCluster = caloClustersEM->emplace_back(
+        emTrackster.regressed_energy(),            // energy
+        math::XYZPoint(emTrackster.barycenter()),  // position
+        reco::CaloID(reco::CaloID::DET_HGCAL_ENDCAP),
+        hitsAndFractions,
+        reco::CaloCluster::particleFlow,  // algoID (copying from output of PFECALCSuperClusterProducer)
+        hitsAndFractions.at(0).first  // seedId : TODO figure out if needed or if we need to put the highest energy hit
+    );
+    caloCluster.setCorrectedEnergy(emTrackster.regressed_energy());
```
Comment (waredjeb) [inline RecoHGCal/TICL/plugins/EGammaSuperclusterProducer.cc:76]: Is it expected to have the regressed energy for the CLUE3D Tracksters? Is the plan to have an energy regression model on the CLUE3D EM Tracksters before the SuperClustering?

File: RecoHGCal/TICL/plugins/TracksterLinkingbySuperClustering.cc [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+  std::stable_sort(
+      trackstersIndicesPt.begin(), trackstersIndicesPt.end(), [&inputTracksters](unsigned int i1, unsigned int i2) {
+        return inputTracksters[i1].raw_pt() > inputTracksters[i2].raw_pt();
+      });
+
+  /* Evaluate in minibatches since running with trackster count = 3000 leads to a short-lived ~15GB memory allocation
+  Also we do not know in advance how many superclustering candidate pairs there are going to be
+  So we pick (arbitrarily) a batch size of 10^5. It needs to be rounded to featureCount
+  */
+  const unsigned int miniBatchSize = static_cast<unsigned int>(1e5) / nnInput->featureCount() * nnInput->featureCount();
```
Comment (waredjeb) [inline RecoHGCal/TICL/plugins/TracksterLinkingbySuperClustering.cc:106]: would it be possible to choose the batch size at configuration?

File: RecoHGCal/TICL/plugins/TrackstersPCA.cc [hunk 3]
```diff
@@ -56,27 +68,71 @@ void ticl::assignPCAtoTracksters(std::vector<Trackster> &tracksters,
...
       barycenter /= trackster.raw_energy();
 
-    // Compute the Covariance Matrix and the sum of the squared weights, used
-    // to compute the correct normalization.
-    // The barycenter has to be known.
-    for (size_t i = 0; i < N; ++i) {
+    LogDebug("TrackstersPCA_Eigen") << "cleaning is  :" << clean << std::endl;
+
+    std::vector<unsigned> filtered_idx;
+    double filtered_energy = 0;
```
Comment (waredjeb) [inline RecoHGCal/TICL/plugins/TrackstersPCA.cc:81]: is it necessary to have double? can we move everything to float? also below

File: RecoHGCal/TICL/plugins/TrackstersPCA.h [hunk 4]
```diff
@@ -4,21 +4,50 @@
 #include "DataFormats/HGCalReco/interface/Trackster.h"
 #include "DataFormats/CaloRecHit/interface/CaloCluster.h"
 #include <vector>
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+
+typedef math::XYZVector Vector;
```
Comment (waredjeb) [inline RecoHGCal/TICL/plugins/TrackstersPCA.h:9]: This XYZVector is a double, is it needed?
If not we could use `ticl::Trackster::Vector`

File: RecoHGCal/TICL/plugins/EGammaSuperclusterProducer.cc (hunk 1 above)
Comment (tcuisset) [reply RecoHGCal/TICL/plugins/EGammaSuperclusterProducer.cc:76]: This still needs to be determined. We need to figure out if the CaloCluster corrected energy is used anywhere in the GsfTrack chain.

File: RecoHGCal/TICL/plugins/TracksterLinkingbySuperClustering.cc (hunk 2 above)
Comment (tcuisset) [reply RecoHGCal/TICL/plugins/TracksterLinkingbySuperClustering.cc]: Done !

File: RecoHGCal/TICL/plugins/TrackstersPCA.cc (hunk 3 above)
Comment (tcuisset) [reply RecoHGCal/TICL/plugins/TrackstersPCA.cc]: I changed to float

File: RecoHGCal/TICL/plugins/TrackstersPCA.h (hunk 4 above)
Comment (tcuisset) [reply RecoHGCal/TICL/plugins/TrackstersPCA.h]: This typedef actually was not used anywhere

**Comment by tcuisset** [comment]
Now rebased against [ticlv5_14_1_X_2024-04-18-2300](https://github.com/felicepantaleo/cmssw/tree/ticlv5_14_1_X_2024-04-18-2300)

**Pull request closed without merging**